governance/src/lib.rs

```diff
@@ -0,0 +1,314 @@
+#![no_std]
+
+pub mod events;
+
+multiversx_sc::imports!();
+multiversx_sc::derive_imports!();
+
+use crate::events::ProposalEventData;
+use gateway::ProxyTrait as _;
+use multiversx_sc::api::KECCAK256_RESULT_LEN;
+
+#[derive(TypeAbi, TopDecode, NestedDecode)]
+pub enum GovernanceCommand {
+    ScheduleTimeLockProposal,
+    CancelTimeLockProposal,
+}
+
+#[derive(TypeAbi, TopDecode)]
+pub struct DecodedCallData<M: ManagedTypeApi> {
+    pub endpoint_name: ManagedBuffer<M>,
+    pub arguments: ManagedVec<M, ManagedBuffer<M>>,
+}
+
+#[derive(TypeAbi, TopDecode)]
+pub struct ExecutePayload<M: ManagedTypeApi> {
+    pub command: GovernanceCommand,
+    pub target: ManagedAddress<M>,
+    pub call_data: ManagedBuffer<M>,
+    pub native_value: BigUint<M>,
+    pub eta: u64,
+}
+
+const EXECUTE_PROPOSAL_CALLBACK_GAS: u64 = 5_000_000; // This is overkill, but the callback should be prevented from failing at all costs
+const KEEP_EXTRA_GAS: u64 = 15_000_000; // Extra gas to keep in contract before registering async promise. This needs to be a somewhat larger value
+
+#[multiversx_sc::contract]
+pub trait Governance: events::Events {
+    #[init]
+    fn init(
+        &self,
+        gateway: ManagedAddress,
+        governance_chain: ManagedBuffer,
+        governance_address: ManagedBuffer,
+        minimum_time_delay: u64,
+    ) {
+        require!(
+            !gateway.is_zero() && !governance_chain.is_empty() && !governance_address.is_empty(),
+            "Invalid address"
+        );
+
+        self.gateway().set(gateway);
+        self.minimum_time_lock_delay()
+            .set(minimum_time_delay);
+
+        self.governance_chain().set(&governance_chain);
+        self.governance_address().set(&governance_address);
+        self.governance_chain_hash()
+            .set(self.crypto().keccak256(&governance_chain));
+        self.governance_address_hash()
+            .set(self.crypto().keccak256(&governance_address));
+    }
+
+    #[upgrade]
+    fn upgrade(&self) {}
+
+    #[payable("EGLD")]
+    #[endpoint(executeProposal)]
+    fn execute_proposal(
+        &self,
+        target: ManagedAddress,
+        call_data: ManagedBuffer,
+        native_value: BigUint,
+    ) {
+        let proposal_hash = self.get_proposal_hash(&target, &call_data, &native_value);
+
+        let eta = self.finalize_time_lock(&proposal_hash);
+
+        self.proposal_executed_event(
+            &proposal_hash,
+            &target,
+            ProposalEventData {
+                call_data: &call_data,
+                value: &native_value,
+            },
+        );
+
+        let decoded_call_data: DecodedCallData<Self::Api> = DecodedCallData::<Self::Api>::top_decode(call_data)
+            .unwrap_or_else(|_| sc_panic!("Could not decode call data"));
+
+        let gas_limit = self.blockchain().get_gas_left() - EXECUTE_PROPOSAL_CALLBACK_GAS - KEEP_EXTRA_GAS;
```

Comment:
Add a check here, for the gas_left to be > EXECUTE_PROPOSAL_CALLBACK_GAS + KEEP_EXTRA_GAS. This way you avoid getting unfriendly VM errors.